presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java

```diff
@@ -123,80 +125,86 @@ private PlanNodeStatsEstimate filterForFalseExpression()
                                             .setDistinctValuesCount(0.0)
                                             .setNullsFraction(NaN).build()));
 
-            return falseStatsBuilder.setOutputRowCount(0.0).build();
+            return Optional.of(falseStatsBuilder.setOutputRowCount(0.0).build());
         }
 
         @Override
-        protected PlanNodeStatsEstimate visitNotExpression(NotExpression node, Void context)
+        protected Optional<PlanNodeStatsEstimate> visitNotExpression(NotExpression node, Void context)
         {
-            return differenceInStats(input, process(node.getValue()));
+            return process(node.getValue()).map(childStats -> differenceInStats(input, childStats));
         }
 
         @Override
-        protected PlanNodeStatsEstimate visitLogicalBinaryExpression(LogicalBinaryExpression node, Void context)
+        protected Optional<PlanNodeStatsEstimate> visitLogicalBinaryExpression(LogicalBinaryExpression node, Void context)
         {
-            PlanNodeStatsEstimate leftStats = process(node.getLeft());
-            PlanNodeStatsEstimate andStats = new FilterExpressionStatsCalculatingVisitor(leftStats, session, types).process(node.getRight());
-
+            Optional<PlanNodeStatsEstimate> leftStats = process(node.getLeft());
+            Optional<PlanNodeStatsEstimate> andStats = new FilterExpressionStatsCalculatingVisitor(leftStats.orElse(filterForUnknownExpression()), session, types).process(node.getRight());
+            if (leftStats.isPresent() && !andStats.isPresent()) {
+                // if we know stats for at least one arg let's use this.
+                andStats = leftStats.map(FilterStatsCalculator::filterStatsForUnknownExpression);
+            }
             switch (node.getType()) {
                 case AND:
                     return andStats;
                 case OR:
-                    PlanNodeStatsEstimate rightStats = process(node.getRight());
-                    PlanNodeStatsEstimate sumStats = addStats(leftStats, rightStats);
-                    return differenceInNonRangeStats(sumStats, andStats);
+                    Optional<PlanNodeStatsEstimate> rightStats = process(node.getRight());
+                    if (!leftStats.isPresent() || !rightStats.isPresent() || !andStats.isPresent()) {
```

Comment:
is it possible for left and right to be present, but not and?  If not, maybe do a checkState for the and instead.